openwisp_controller/connection/connectors/snmp.py

```diff
@@ -0,0 +1,51 @@
+import logging
+
+from django.utils.functional import cached_property
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError as SchemaError
+
+logger = logging.getLogger(__name__)
+
+
+class Snmp(object):
+    schema = {
+        '$schema': 'http://json-schema.org/draft-04/schema#',
+        'type': 'object',
+        'title': 'Credentials type',
+        'oneOf': [
+            {
+                'title': 'SNMP',
+                'required': ['community', 'agent'],
+                'additionalProperties': False,
+                'properties': {
+                    'community': {'type': 'string', 'default': 'public'},
+                    'agent': {'type': 'string'},
+                    'port': {
+                        'type': 'integer',
+                        'default': 161,
+                        'minimum': 1,
+                        'maximum': 65535,
+                    },
+                },
+            },
+        ],
+    }
```

Comment:
Can you please explain the requirement of using 'OneOf' here?

Reply:
@pandafy I'm not sure of a suitable alternative. Should we use `allOf` or something else?

Reply:
```python
{
        '$schema': 'http://json-schema.org/draft-04/schema#',
        'type': 'object',
        'title': 'SNMP',
        'required': ['community', 'agent'],
        'additionalProperties': False,
        'properties': {
            'community': {'type': 'string', 'default': 'public'},
            'agent': {'type': 'string'},
            'port': {
                'type': 'integer',
                'default': 161,
                'minimum': 1,
                'maximum': 65535,
            },
        },
    }
 ```
 I tried above schema and it worked. `allOf` or `oneOf` are useful if there can be multiple options. For example, the SSH credential can either have a username and password or a username and an SSH key. Hence, `oneOf` is used for it's schema. 
 
 Let's wait for @nemesisdesign confirmation before making this change.

Reply:
oneOf is good because it creates a select which allows to change to other one type or the other.
I think we could use something similar here if there are multiple ways of specifying authentication and every way has a set of well defined fields.

As @pandafy said, with SSH, one can login with either username and password or username and private key, so the schema reflects that, it does not allow you to insert username, password and private key because it doesn't make sense.

@purhan shouldn't we allow users to define the SNMP version in use? Isn't this necessary when connecting to devices?

Reply:
> @purhan shouldn't we allow users to define the SNMP version in use? Isn't this necessary when connecting to devices?

@nemesisdesign We don't define a version anywhere in `pysnmp`. So there is no way we can use it in Netengine and don't need it here